File: lib/JFFStoreKit/JFFStoreKit/asyncAdapters/asyncSKPaymentQueue.m
#import "asyncSKPaymentQueue.h"

#import "JFFStoreKitDisabledError.h"
#import "JFFStoreKitTransactionStateFailedError.h"

#import "asyncSKFinishTransaction.h"

@interface JFFAsyncSKPaymentAdapter : NSObject <
SKPaymentTransactionObserver,
JFFAsyncOperationInterface
>
@end

@implementation JFFAsyncSKPaymentAdapter
{
    SKPaymentQueue *_queue;
    SKPayment *_payment;
    BOOL _addedToObservers;
    JFFDidFinishAsyncOperationCallback _finishCallback;
}

- (void)dealloc
{
    [self unsubscribeFromObservervation];
    _finishCallback = nil;
}

- (void)doNothing:(id)objetc
{
}

- (void)unsubscribeFromObservervation
{
    if (_addedToObservers) {
        [_queue removeTransactionObserver:self];
        _addedToObservers = NO;
    }
}

+ (instancetype)newAsyncSKPaymentAdapterWithRequest:(SKPayment *)payment
{
    JFFAsyncSKPaymentAdapter *result = [self new];
    
    if (result) {
        result->_payment = payment;
        result->_queue   = [SKPaymentQueue defaultQueue];
        
        [result->_queue addTransactionObserver:result];
        result->_addedToObservers = YES;
    }
    
    return result;
}

- (void)asyncOperationWithResultCallback:(JFFDidFinishAsyncOperationCallback)finishCallback
                         handlerCallback:(JFFAsyncOperationChangeStateCallback)handlerCallback
                        progressCallback:(JFFAsyncOperationProgressCallback)progressCallback
{
    if (![SKPaymentQueue canMakePayments]) {
        finishCallback(nil, [JFFStoreKitDisabledError new]);
        return;
    }
    
    _finishCallback = [finishCallback copy];
    
    SKPaymentTransaction *transaction = [self ownPurchasedTransaction];
    
    if (transaction) {
        
        [self unsubscribeFromObservervation];
        _finishCallback(transaction, nil);
    } else {
        
        [_queue addPayment:_payment];
    }
}

- (void)doTask:(JFFAsyncOperationHandlerTask)task
{
    NSParameterAssert(task <= JFFAsyncOperationHandlerTaskCancel);
    
    if (task == JFFAsyncOperationHandlerTaskUnSubscribe)
        [self unsubscribeFromObservervation];
}

- (SKPaymentTransaction *)ownPurchasedTransaction
{
    //SKPayment
    SKPaymentTransaction *transaction = [_queue.transactions firstMatch:^BOOL(SKPaymentTransaction *transaction) {
        
        return transaction.transactionState == SKPaymentTransactionStatePurchased
        && [_payment.productIdentifier isEqualToString:transaction.payment.productIdentifier];
    }];
    
    return transaction;
}

- (SKPaymentTransaction *)ownTransactionForTransactions:(NSArray *)transactions
{
    SKPaymentTransaction *transaction = [transactions firstMatch:^BOOL(SKPaymentTransaction *transaction) {
        return [_payment isEqual:transaction.payment];
    }];
    
    return transaction;
}

#pragma mark SKPaymentTransactionObserver

- (void)paymentQueue:(SKPaymentQueue *)queue updatedTransactions:(NSArray *)transactions
{
    if (!_finishCallback) {
        return;
    }
    
    SKPaymentTransaction *transaction = [self ownTransactionForTransactions:transactions];
    
    if (!transaction) {
        return;
    }
    
    //TODO fix workaround for IOS 6.0
    [self performSelector:@selector(doNothing:) withObject:self afterDelay:1.];
    
    switch (transaction.transactionState)
    {
        case SKPaymentTransactionStatePurchased:
        {
            [self unsubscribeFromObservervation];
            _finishCallback(transaction, nil);
            break;
        }
        case SKPaymentTransactionStateFailed:
        {
            if (transaction.error.code != SKErrorPaymentCancelled) {
                // Optionally, display an error here.
            }
            JFFStoreKitTransactionStateFailedError *error = [JFFStoreKitTransactionStateFailedError new];
            error.transaction = transaction;
            [self unsubscribeFromObservervation];
            _finishCallback(nil, error);
            break;
        }
        case SKPaymentTransactionStateRestored:
        {
            [self unsubscribeFromObservervation];
            _finishCallback(transaction, nil);
            break;
        }
        default:
            break;
    }
    // TODO call progress with SKPaymentTransactionStatePurchasing
}

@end

JFFAsyncOperation asyncOperationWithSKPayment(SKPayment *payment)
{
    JFFAsyncOperationInstanceBuilder factory = ^id< JFFAsyncOperationInterface >() {
        return [JFFAsyncSKPaymentAdapter newAsyncSKPaymentAdapterWithRequest:payment];
    };
    JFFAsyncOperation loader = buildAsyncOperationWithAdapterFactory(factory);
    
    loader = bindTrySequenceOfAsyncOperations(loader, ^JFFAsyncOperation(JFFStoreKitTransactionStateFailedError *error) {
        
        if (![error isKindOfClass:[JFFStoreKitTransactionStateFailedError class]]) {
            
            return asyncOperationWithError(error);
        }
        
        JFFAsyncOperation loader = trySequenceOfAsyncOperations(asyncOperationFinishTransaction(error.transaction), asyncOperationWithResult(@YES), nil);
        
        return sequenceOfAsyncOperations(loader, asyncOperationWithError(error), nil);
    }, nil);
    
    return loader;
}

File: iAsync.Swift/lib/JStoreKit/JStoreKit/asyncAdapters/asyncSKPaymentQueue.swift
//
//  asyncSKPaymentQueue.swift
//  JStoreKit
//
//  Created by Vladimir Gorbenko on 29.09.14.
//  Copyright (c) 2014 EmbeddedSources. All rights reserved.
//

import Foundation

import JAsync
import JUtils
import StoreKit
import CoreFoundation

//TODO private !!!
public class JAsyncSKPaymentAdapter : NSObject, SKPaymentTransactionObserver, JAsyncInterface {
    
    public typealias ResultType = SKPaymentTransaction
    
    private let queue  : SKPaymentQueue
    private let payment: SKPayment
    private var addedToObservers = true
    private var finishCallback: JAsyncTypes<ResultType>.JDidFinishAsyncCallback?
    
    deinit {
        unsubscribeFromObservervation()
        finishCallback = nil
    }
    
    @objc(doNothing:)//TODO remove
    public func doNothing(objetc: AnyObject) {}
    
    func unsubscribeFromObservervation() {
        
        if addedToObservers {
            queue.removeTransactionObserver(self)
            addedToObservers = false
        }
    }
    
    init(payment: SKPayment) {
        
        self.payment = payment
        self.queue   = SKPaymentQueue.defaultQueue()
        
        super.init()
        
        self.queue.addTransactionObserver(self)
    }
    
    public var isForeignThreadResultCallback: Bool {
        return true
    }

    public func asyncWithResultCallback(
        finishCallback: JAsyncTypes<ResultType>.JDidFinishAsyncCallback,
        stateCallback: JAsyncChangeStateCallback,
        progressCallback: JAsyncProgressCallback)
    {
        if !SKPaymentQueue.canMakePayments() {
            finishCallback(result: JResult.error(JStoreKitDisabledError()))
            return
        }
    
        self.finishCallback = finishCallback
        queue.addPayment(payment)
    }
    
    public func doTask(task: JAsyncHandlerTask)
    {
        assert(task.rawValue <= JAsyncHandlerTask.Cancel.rawValue)
    
        if task == .UnSubscribe {
            unsubscribeFromObservervation()
        }
    }
    
    func ownTransactionForTransactions(transactions: [SKPaymentTransaction]) -> SKPaymentTransaction?
    {
        let transaction = firstMatch(transactions) { (transaction: SKPaymentTransaction) -> Bool in
            return self.payment.isEqual(transaction.payment)
        }
        
        return transaction
    }

    public func paymentQueue(queue: SKPaymentQueue!, updatedTransactions transactions: [AnyObject]!) {
        
        if self.finishCallback == nil {
            return
        }
        
        let transactionsAr = transactions as! [SKPaymentTransaction]
    
        let transaction = ownTransactionForTransactions(transactionsAr)
    
        if let transaction = transaction {
            
            //TODO fix workaround for IOS 6.0
            NSTimer.scheduledTimerWithTimeInterval(1.1, target: self, selector: Selector("doNothing:"), userInfo: nil, repeats: false)
            
            switch transaction.transactionState {
                
            case SKPaymentTransactionState.Purchased:
                unsubscribeFromObservervation()
                finishCallback?(result: JResult.value(transaction))
            case SKPaymentTransactionState.Failed:
                var error = { () -> NSError in
                    
                    if SKErrorDomain == transaction.error.domain
                        && transaction.error.code == SKErrorPaymentCancelled {
                        
                        return JAsyncFinishedByCancellationError()
                    }
                    
                    let locError = JStoreKitTransactionStateFailedError(transaction:transaction)
                    return locError
                }()
                unsubscribeFromObservervation()
                finishCallback?(result: JResult.error(error))
            case SKPaymentTransactionState.Restored:
                unsubscribeFromObservervation()
                finishCallback?(result: JResult.value(transaction))
            ////TODO !!! process SKPaymentTransactionStateDeferred also
            default:
                break
            }
        }
        // TODO call progress with SKPaymentTransactionStatePurchasing
    }
}

public func asyncWithSKPayment(payment: SKPayment) -> JAsyncTypes<SKPaymentTransaction>.JAsync
{
    let factory = { () -> JAsyncSKPaymentAdapter in
        return JAsyncSKPaymentAdapter(payment: payment)
    }
    var loader = JAsyncBuilder.buildWithAdapterFactory(factory)
    
    loader = bindTrySequenceOfAsyncs(loader, { (error: NSError) -> JAsyncTypes<SKPaymentTransaction>.JAsync in
        
        if let error = error as? JStoreKitTransactionStateFailedError {
            
            let loader = trySequenceOfAsyncs(asyncFinishTransaction(error.transaction), asyncWithResult([]))
            
            let errorLoader: JAsyncTypes<SKPaymentTransaction>.JAsync = asyncWithError(error)
            return sequenceOfAsyncs(loader, errorLoader)
        }
        
        return asyncWithError(error)
    })
    
    return loader
}
